rust/src/composepost.rs

```diff
@@ -202,6 +202,47 @@ fn postprocess_presets(rootfs_dfd: &Dir) -> Result<()> {
     Ok(())
 }
 
+fn is_overlay_whiteout(meta: &cap_std::fs::Metadata) -> bool {
+    (meta.mode() & libc::S_IFMT) == libc::S_IFCHR && meta.rdev() == 0
+}
+
+/// Auto-synthesize embedded overlayfs whiteouts; for more information
+/// see https://github.com/ostreedev/ostree/pull/2722/commits/0085494e350c72599fc5c0e00422885d80b3c660
+#[context("Postprocessing embedded overlayfs")]
+fn postprocess_embedded_ovl_whiteouts(root: &Dir) -> Result<()> {
+    const OSTREE_WHITEOUT_PREFIX: &str = ".ostree-wh.";
+    fn recurse(root: &Dir, path: &Utf8Path) -> Result<u32> {
+        let mut n = 0;
+        for entry in root.read_dir(path)? {
+            let entry = entry?;
+            let meta = entry.metadata()?;
+            let name = PathBuf::from(entry.file_name());
+            let name: Utf8PathBuf = name.try_into()?;
+            if meta.is_dir() {
+                n += recurse(root, &path.join(name))?;
+                continue;
+            }
+            if !is_overlay_whiteout(&meta) {
+                continue;
+            };
+            let srcpath = path.join(&name);
+            let targetname = format!("{OSTREE_WHITEOUT_PREFIX}{name}");
+            let destpath = path.join(&targetname);
```

Comment:
I haven't seen any effect on overlayfs actually, on extraction I just added it as caution,
but should we try to preserve permissions/xattrs?

Reply:
On permissions: Hmm...ISTM that the standard for these whiteouts is to be mode 000, so why would we support anything else?  We could just forcibly set that mode on the ostree side in at checkout time, instead of requiring that permission to be on the regular file in the tree.

As a side note, files shipped with no permissions, not even user-read are *really annoying* because they trip up doing things without privileges.  It only works when run as root (or really, `CAP_DAC_OVERRIDE`) which you can get in a user namespace, but still.  We've hit corner cases around this dealing with `/etc/shadow` shipped in an ostree.

Anyways so...for now yeah I can set this to permissions 0, but I'd argue we should revisit this and have the ostree side force on permissions 0 and ignore whatever the source file permissions are.

On xattrs: Note that ostree handles selinux labeling (`security.selinux`) itself at commit time.  I can't think of any other xattrs that could be relevant here.  There's IMA but that only supports regular files, and in the end the whiteout file is interpreted by the `overlayfs` kernel module directly, it doesn't have any contents.  It surely doesn't make sense to add e.g. file capabilities to a whiteout...

Reply:
Updated to propagate permissions.